src/Libraries/Microsoft.Extensions.AI/ToolReduction/EmbeddingToolReductionStrategy.cs

```diff
@@ -0,0 +1,212 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Numerics.Tensors;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+#pragma warning disable IDE0032 // Use auto property, suppressed until repo updates to C# 14
+
+/// <summary>
+/// A tool reduction strategy that ranks tools by embedding similarity to the current conversation context.
+/// </summary>
+/// <remarks>
+/// The strategy embeds each tool (name + description by default) once (cached) and embeds the current
+/// conversation content each request. It then selects the top <c>toolLimit</c> tools by similarity.
+/// </remarks>
+[Experimental("MEAI001")]
+public sealed class EmbeddingToolReductionStrategy : IToolReductionStrategy
+{
+    private readonly ConditionalWeakTable<AITool, Embedding<float>> _toolEmbeddingsCache = new();
+    private readonly IEmbeddingGenerator<string, Embedding<float>> _embeddingGenerator;
+    private readonly int _toolLimit;
+
+    private Func<AITool, string> _toolEmbeddingTextFactory = static t =>
+    {
+        if (string.IsNullOrWhiteSpace(t.Name))
+        {
+            return t.Description;
+        }
+
+        if (string.IsNullOrWhiteSpace(t.Description))
+        {
+            return t.Name;
+        }
+
+        return t.Name + "\n" + t.Description;
+    };
+
+    private Func<IEnumerable<ChatMessage>, string> _messagesEmbeddingTextFactory = static messages =>
+    {
+        var messageTexts = messages.Select(m => m.Text).Where(s => !string.IsNullOrEmpty(s));
+        return string.Join("\n", messageTexts);
```

Comment:
We have a helper that ChatMessage and ChatResponse use to do this a bit more efficiently. Maybe we should look at exposing that in a way this could use as well.

That said, I wonder if we'd want the default implementation here also including reasoning content, which this implementation doesn't currently, since ChatMessage.Text only includes TextContent and not TextReasoningContent.